care/emr/utils/decision_engine/base.py

```diff
@@ -0,0 +1,301 @@
+from collections.abc import Callable
+from typing import Any
+
+
+class AbstractRuleEngine:
+    """
+    An abstract base class for rule engines that evaluate conditions using registered handlers and validators.
+    """
+
+    def _register_internal_handlers(self):
+        """
+        Register default handler functions for condition evaluation.
+
+        This method registers built-in handlers like equality, range, in_or_equality, and in.
+        It ensures handlers are only added if not already present.
+        """
+        handlers = {
+            "equality": self._handle_equality,
+            "range": self._handle_range,
+            "in_or_equality": self._handle_in_or_equality,
+            "in": self._handle_in,
+        }
+        for name, func in handlers.items():
+            if name not in self._handler_registry:
+                self._handler_registry[name] = func
+
+    def _register_internal_validators(self):
+        """
+        Register default validator functions for condition specs.
+
+        This method registers built-in validators corresponding to the internal handlers.
+        It ensures validators are only added if not already present.
+        """
+        validators = {
+            "equality": lambda expected_value: isinstance(
+                expected_value, (str, int, float)
+            ),
+            "range": lambda expected_value: isinstance(expected_value, dict)
+            and all(
+                expected_value.get(key) is None
+                or isinstance(expected_value.get(key), (int, float))
+                for key in ["min", "max"]
+                if key in expected_value
+            ),
+            "in_or_equality": lambda expected_value: isinstance(
+                expected_value, (str, list)
+            ),
+            "in": lambda expected_value: isinstance(expected_value, list),
+        }
+        for name, func in validators.items():
+            if name not in self._validator_registry:
+                self._validator_registry[name] = func
+
+    def register_internal_conditions(self):
+        """
+        Register internal condition mappings (e.g., gender to equality).
+
+        Raises:
+            NotImplementedError: Subclasses must implement this method to define internal condition mappings.
+        """
+        raise NotImplementedError(
+            "Subclasses must implement register_internal_conditions"
+        )
+
```

Comment:
we shouldn't be raising an error here, this might break engines that dont override this handler